vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,917 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "automl.h"
+#include "debug_log.h"
+#include "reductions.h"
+#include "learner.h"
+#include "io/logger.h"
+#include "vw.h"
+
+#include <cfloat>
+
+using namespace VW::config;
+using namespace VW::LEARNER;
+
+namespace logger = VW::io::logger;
+
+/*
+This reduction implements the ChaCha algorithm from page 5 of the following paper:
+https://arxiv.org/pdf/2106.04815.pdf
+
+There are two key differences between this implementation and the algorithm described
+in the paper. First, the paper assumes all knowledge (of examples and namespaces) is
+known at the start, whereas that information is gathered as we process examples in reality.
+Second, this algorithm follows the pattern Schedule -> Update Champ -> Learn, whereas
+the paper follows the pattern Schedule -> Learn -> Update Champ. This should not make a
+functional difference. Nearly all of the variables and functions described in the
+algorithm live in the config_manager struct. The following provides a translation of
+terms used in the paper to this implementation:
+
+Variables from ChaCha:
+c_init -> 0: The first champ is initialized to the 0th indexed set of weights
+b(budget) -> max_live_configs: The weights/configs to learn on
+C -> current_champ: The weight index of the current champ
+S -> configs: The entire set of challengers/configs generated so far
+B(+ C) -> scores: The live challengers (including the champ) with statistics
+
+Functions from ChaCha:
+ConfigOracle -> exclusion_configs_oracle(): Generates new configs based on champ
+Schedule -> schedule(): Swaps out / manages live configs on each step
+Schedule.median -> better_than_median(): Checks if config is better than median
+Schedule.Choose -> calc_priority(): Determines priority when selecting next live config
+Predict/Incur Loss -> offset_learn(): Updates weights and learns on live configs
+Better -> better(): Changes champ if challenger is better
+Worse -> worse(): Removes challenger is much worse than champ
+*/
+
+namespace VW
+{
+namespace automl
+{
+namespace details
```

Comment:
`details` as in implementation details? Can this namesepace be more specific? e.g. `internals` or something

Reply:
This was mentioned in a previous comment and we changed it from `helper` to `details` for consistency with other reductions. The scope of this namespace is only the `fail_if_enabled` function, which should be removed eventually, so this point may be moot.

Reply:
ok